### PR TITLE
Add option to bake a mesh from animated skeleton pose

### PR DESCRIPTION
**Pull request by smix8**
Adds option to bake a mesh from animated skeleton  pose.

Takes a snapshot of the current animated skeleton pose of the skinned mesh and bakes it to another ArrayMesh.

Implements https://github.com/godotengine/godot-proposals/issues/8120 to make stuff like this:

skinned_mesh_baking

Pairs with https://github.com/godotengine/godot/pull/76725 to bake the blend shapes.

_Production edit: closes godotengine/godot-roadmap#58_

### CODE REVIEW

**Comment by jcostello** [comment]
Could be use to emit particles from a animated mesh?

**Comment by smix8** [comment]
> Could be use to emit particles from a animated mesh?

See https://github.com/godotengine/godot-proposals/issues/8106 and https://github.com/godotengine/godot-proposals/issues/5051

Particles have currently no exposed API to bake emission points from mesh at runtime.

For performance it also makes far more sense to directly bake to the texture that the particle shader uses instead of baking to an intermediate mesh.

**Comment by clayjohn** [comment]
Personally I am a bit wary of exposing new APIs to users that cause GPU stalls like this. My preference would be to first add a caching system to Mesh so that users can choose to keep the data on the CPU if they want to do things like this.

**Comment by smix8** [comment]
Where should such an option to (pre)cache the mesh arrays be added? On the `MeshInstance3D` or the `Mesh` resource?

**Review by fire — APPROVED** [review]
I am in favour of the feature, but I don't know how to resolve the request from @clayjohn

**Comment by fire** [comment]
I would also reuse the same code from the blend shape bake with the inputted mesh

**Comment by smix8** [comment]
Applied the same changes as in the blendshape bake pr. The baked mesh is always returned, and if an optional existing mesh param is provided it updates this mesh RID instead of creating a new mesh as this helps with updating instances that already use that mesh.

File: doc/classes/MeshInstance3D.xml [hunk 1]
```diff
@@ -13,6 +13,14 @@
 		<link title="Third Person Shooter (TPS) Demo">https://godotengine.org/asset-library/asset/2710</link>
 	</tutorials>
 	<methods>
+		<method name="bake_mesh_from_current_skeleton_pose">
+			<return type="ArrayMesh" />
+			<param index="0" name="existing" type="ArrayMesh" default="null" />
+			<description>
+				Takes a snapshot of the current animated skeleton pose of the skinned mesh and bakes it to the provided [param existing] mesh. If no [param existing] mesh is provided a new [ArrayMesh] is created, baked and returned. Requires a skeleton with registered skin to work. Blendshapes are ignored. Mesh surface materials are not copied.
+				[b]Performance:[/b] [Mesh] data needs to be received from the GPU, stalling the [RenderingServer] in the process.
```
Comment (AThousandShips) [inline doc/classes/MeshInstance3D.xml:21]:
```suggestion
				[b]Performance:[/b] [Mesh] data needs to be retrieved from the GPU, stalling the [RenderingServer] in the process.
```

File: scene/3d/mesh_instance_3d.cpp [hunk 2]
```diff
@@ -515,6 +515,108 @@ bool MeshInstance3D::_property_get_revert(const StringName &p_name, Variant &r_p
...
+
+		unsigned int vertex_count = source_mesh_vertex_array.size();
+		int expected_bone_array_size = vertex_count * bones_per_vertex;
+		ERR_CONTINUE(source_mesh_bones_array.size() != expected_bone_array_size);
+		ERR_CONTINUE(source_mesh_weights_array.size() != expected_bone_array_size);
+
+		Array new_mesh_arrays;
+		new_mesh_arrays.resize(Mesh::ARRAY_MAX);
+		for (int i = 0; i < source_mesh_arrays.size(); i++) {
+			if (i == Mesh::ARRAY_VERTEX || i == Mesh::ARRAY_BONES || i == Mesh::ARRAY_WEIGHTS) {
```
Comment (caphindsight) [inline scene/3d/mesh_instance_3d.cpp:582]: Should the normals not be rotated with the bones?

Comment (smix8) [reply scene/3d/mesh_instance_3d.cpp]: added

**Comment by PiCode9560** [comment]
Is this thing done yet?

**Review by fire — APPROVED** [review]

**Review by AThousandShips — DISMISSED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by RavinMaddHatter** [comment]
Does this actually work? i have been trying to get it to work for a few days. Any example code would be nice or an example project. i tried what is in the docs but i keep getting out the orginal unblended mesh.

**Comment by fire** [comment]
@RavinMaddHatter are you testing on 4.4?

**Comment by RavinMaddHatter** [comment]
image
I have tested on 4.4 Dev2 I have tried many many things. I am very likely doing something wrong, but that would mean the documentation is incomplete (happy to help fix it) or there is a bug. If i can get you something for docs update if we can figure out how to get it to work on my PC.

Can chat in discord or continue here. I have several posts with no responses.

**Comment by RavinMaddHatter** [comment]
Does anyone have a test project so i can figure out the difference and why i only get the default pose?

**Comment by qualmonade** [comment]
I'm currently encountering a problem where regardless of if it's true or not, the function does not seem to find any of the bones attached to the skeleton of my mesh of choice. I have so far tested this with inherited scenes, including imported GLB files.
image
image
image

I am wondering if this is related to the issue that RavinMaddHatter is having?
> Does this actually work? i have been trying to get it to work for a few days. Any example code would be nice or an example project. i tried what is in the docs but i keep getting out the orginal unblended mesh.

**Comment by fire** [comment]
Please provide a test project with code.

**Comment by qualmonade** [comment]
[Here you go!](https://github.com/qualmonade/Bake_Mesh_From_Pose-Test)
I am hoping this is correct. Let me know if anything is broken in some way or other.

**Comment by RavinMaddHatter** [comment]
This may be my issue, I will test when I can.

**Comment by RavinMaddHatter** [comment]
Looking at this again. This is the node i am baking. It is generated via another library so it is called Baked_Opaque. 
image

In that node, there is an attached skeleton:
image
I am using this exact line of code to get the meshinstance3d baked from the skeleton
```var mesh = humanizer.find_child("Baked-Opaque").bake_mesh_from_current_blend_shape_mix()```
Using this version of godot
image
When i export the OBJ (i have been exporting OBJ files of the returned mesh instance 3d rather than just displaying it, because this is supposed to return the mesh snap shot, not just a copy of the skeleton attached to a similar mesh)

This is the posed input mesh rendered by the GPU
image
This is the mesh exported from the CPU
image

I am happy to be wrong here, just want to know where i am wrong.

**Comment by RavinMaddHatter** [comment]
You can see the project. it is not minimized i apologize for that, but I am not sure how to minimize it right now.  I am still figuring out the 3d aspects of godot. 
https://github.com/RavinMaddHatter/Madhatters-Table-Top-RPG-Mini-Maker
The baking is occuring on line 439 of new_character_menu script
The export is occurring on line 445

This is a learning project so the code hygiene is terrible. But i cannot get it to work, i am looking into other options on this topic as well.

**Comment by fire** [comment]
I have made an issue on https://github.com/RavinMaddHatter/Madhatters-Table-Top-RPG-Mini-Maker to debug this.

**Comment by RavinMaddHatter** [comment]
From the other thread "There seems to be a missing feature to bake both blend shapes and skeleton poses."

I have not seen any updates. but i am curious if this is a bug in my implementation or in the godot implementation.

**Pull request merged**